[CI] MultiFeatureMigrationIT testMultipleFeatureMigration failing

**Build scan:**
https://gradle-enterprise.elastic.co/s/awf556kh43gjw/tests/:modules:reindex:internalClusterTest/org.elasticsearch.migration.MultiFeatureMigrationIT/testMultipleFeatureMigration

**Reproduction line:**
`./gradlew ':modules:reindex:internalClusterTest' --tests "org.elasticsearch.migration.MultiFeatureMigrationIT.testMultipleFeatureMigration" -Dtests.seed=79FBE8C22EDD086A -Dtests.locale=ar-KW -Dtests.timezone=Asia/Macau -Druntime.java=17`

**Applicable branches:**
8.0

**Reproduces locally?:**
Yes

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.migration.MultiFeatureMigrationIT&tests.test=testMultipleFeatureMigration

**Failure excerpt:**
```
java.lang.AssertionError: 
Expected: <NO_MIGRATION_NEEDED>
     but: was <IN_PROGRESS>

  at __randomizedtesting.SeedInfo.seed([79FBE8C22EDD086A:9BB132EC51E05DC5]:0)
  at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
  at org.junit.Assert.assertThat(Assert.java:956)
  at org.junit.Assert.assertThat(Assert.java:923)
  at org.elasticsearch.migration.MultiFeatureMigrationIT.lambda$testMultipleFeatureMigration$4(MultiFeatureMigrationIT.java:185)
  at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1069)
  at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1042)
  at org.elasticsearch.migration.MultiFeatureMigrationIT.testMultipleFeatureMigration(MultiFeatureMigrationIT.java:181)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:568)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:824)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:475)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:831)
  at java.lang.Thread.run(Thread.java:833)

```

--- a/modules/reindex/src/internalClusterTest/java/org/elasticsearch/migration/MultiFeatureMigrationIT.java
+++ b/modules/reindex/src/internalClusterTest/java/org/elasticsearch/migration/MultiFeatureMigrationIT.java
@@ -37,6 +37,8 @@ import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.BiConsumer;
 import java.util.function.Function;
@@ -91,6 +93,8 @@ public class MultiFeatureMigrationIT extends FeatureMigrationIT {
 
         ensureGreen();
 
+        CountDownLatch hooksCalled = new CountDownLatch(4);
+
         SetOnce<Boolean> preMigrationHookCalled = new SetOnce<>();
         SetOnce<Boolean> postMigrationHookCalled = new SetOnce<>();
         SetOnce<Boolean> secondPluginPreMigrationHookCalled = new SetOnce<>();
@@ -109,6 +113,7 @@ public class MultiFeatureMigrationIT extends FeatureMigrationIT {
             assertThat(currentResults, nullValue());
 
             preMigrationHookCalled.set(true);
+            hooksCalled.countDown();
             return metadata;
         });
 
@@ -125,6 +130,7 @@ public class MultiFeatureMigrationIT extends FeatureMigrationIT {
             assertThat(currentResults, nullValue());
 
             postMigrationHookCalled.set(true);
+            hooksCalled.countDown();
         });
 
         SecondPlugin.preMigrationHook.set(clusterState -> {
@@ -145,6 +151,7 @@ public class MultiFeatureMigrationIT extends FeatureMigrationIT {
             assertThat(currentResults.getFeatureStatuses().get(FEATURE_NAME).getException(), nullValue());
 
             secondPluginPreMigrationHookCalled.set(true);
+            hooksCalled.countDown();
             return metadata;
         });
 
@@ -165,6 +172,7 @@ public class MultiFeatureMigrationIT extends FeatureMigrationIT {
             assertThat(currentResults.getFeatureStatuses().get(FEATURE_NAME).getException(), nullValue());
 
             secondPluginPostMigrationHookCalled.set(true);
+            hooksCalled.countDown();
         });
 
         PostFeatureUpgradeRequest migrationRequest = new PostFeatureUpgradeRequest();
@@ -176,6 +184,9 @@ public class MultiFeatureMigrationIT extends FeatureMigrationIT {
             .map(PostFeatureUpgradeResponse.Feature::getFeatureName)
             .collect(Collectors.toSet());
         assertThat(migratingFeatures, hasItems(FEATURE_NAME, SECOND_FEATURE_NAME));
+
+        // wait for all the plugin methods to have been called before assertBusy since that will exponentially backoff
+        assertThat(hooksCalled.await(30, TimeUnit.SECONDS), is(true));
 
         GetFeatureUpgradeStatusRequest getStatusRequest = new GetFeatureUpgradeStatusRequest();
         assertBusy(() -> {

